Backport 2.x: Add missing PSA macros

gilles-peskine-arm opened the pull request:
Backport of https://github.com/ARMmbed/mbedtls/pull/4966

Differences with the 3.x version:

* Add/add conflict in "Add missing PSA_ALG_NONE macro."
* Update stored generated file in "Add missing PSA_ALG_NONE macro."
* Add support for MD2 and MD4 in "Add missing PSA_HASH_BLOCK_LENGTH macro."
* Remove the MD2/MD4 lines as well in "Use the new macro PSA_HASH_BLOCK_LENGTH"

mpg reviewed (dismissed):
Approving as a faithful backport of #4966 with the differences mentioned in the description, which are all justified.

gabor-mezei-arm reviewed (approved):
Looks good to me

mpg commented:
This now has conflicts and needs rebasing.

gilles-peskine-arm commented:
I rebased due to a merge conflict due to interactions with #5010/#5117 (new PSS variant). The previous version ishttps://github.com/gilles-peskine-arm/mbedtls/tree/missing-psa-macros-2.x-1 . Rebase notes:

* "Untangle PSA_ALG_IS_HASH_AND_SIGN and PSA_ALG_IS_SIGN_HASH": this commit replaces `ALG_IS_HASH_AND_SIGN` by `ALG_IS_SIGN_HASH`. Do it in the places in `tests/suites/test_suite_psa_crypto_metadata.data` and `tests/suites/test_suite_psa_crypto_metadata.function` affected by the changes for the new PSS variant.
* "Add PSA_ALG_IS_HASH_AND_SIGN to the metadata tests": do it for the two PSS test cases in `tests/suites/test_suite_psa_crypto_metadata.data` affected by the changes for the new PSS variant. In `tests/suites/test_suite_psa_crypto_metadata.function`, both this commit and the PSS PR added entries to the list of classification macros, so modify this commit due to the numbers being shifted.

While rebasing, I noticed a preexisting bug in the test code that led to an unintended test gap. I added two new commits: "Ensure that all flags are actually tested" is a non-regression test and "Fix test bug: some classification flags were not tested" fixes the bug.

mpg reviewed (approved):
I've reviewer the rebase and the two new commits, LGTM.

gabor-mezei-arm reviewed (approved):
Looks good to me
